src/lib/hooks/routing/clientSideSmartOrderRouter.ts

```diff
@@ -1,25 +1,47 @@
+import type { Web3Provider } from '@ethersproject/providers'
 import { BigintIsh, ChainId, CurrencyAmount, Token, TradeType } from '@uniswap/sdk-core'
 // This file is lazy-loaded, so the import of smart-order-router is intentional.
 // eslint-disable-next-line @typescript-eslint/no-restricted-imports
 import { AlphaRouter, AlphaRouterConfig } from '@uniswap/smart-order-router'
 import { asSupportedChain } from 'constants/chains'
-import { DEPRECATED_RPC_PROVIDERS } from 'constants/providers'
+import { RPC_PROVIDERS } from 'constants/providers'
 import { nativeOnChain } from 'constants/tokens'
 import JSBI from 'jsbi'
+import AppStaticJsonRpcProvider from 'rpc/StaticJsonRpcProvider'
 import { GetQuoteArgs, QuoteResult, QuoteState, SwapRouterNativeAssets } from 'state/routing/types'
 import { transformSwapRouteToGetQuoteResult } from 'utils/transformSwapRouteToGetQuoteResult'
 
-const routers = new Map<ChainId, AlphaRouter>()
-export function getRouter(chainId: ChainId): AlphaRouter {
+type RouterAndProvider = { router: AlphaRouter; provider: AppStaticJsonRpcProvider | Web3Provider }
+let cachedProviderRouter: { chainId: number; routerProvider: RouterAndProvider } | undefined = undefined
+const routers = new Map<ChainId, RouterAndProvider>()
+export function getRouter(chainId: ChainId, web3Provider: Web3Provider | undefined): RouterAndProvider {
+  const providerChainId = web3Provider?.network.chainId
+  if (
+    cachedProviderRouter !== undefined &&
+    cachedProviderRouter.chainId === providerChainId &&
+    chainId === providerChainId &&
+    web3Provider === cachedProviderRouter.routerProvider.provider
+  ) {
+    return cachedProviderRouter.routerProvider
+  } else {
+    cachedProviderRouter = undefined
+  }
+  if (providerChainId !== undefined && chainId === providerChainId && web3Provider !== undefined) {
```

Comment:
Last check is redundant.  If `providerChainId` is defined then `web3Provider` will never be `undefined`.  I'm a bit sad that TypeScript doesn't notice this.
```suggestion
  if (providerChainId !== undefined && chainId === providerChainId) {
```
But maybe you need the check in order for TS to properly narrow?

Reply:
its needed for typescript